node/src/components/storage.rs

```diff
@@ -1182,14 +1182,27 @@ impl Storage {
         Ok(maybe_block)
     }
 
-    pub(crate) fn read_blocks_since(
+    /// Retrieves the contiguous segment of the block chain starting at the highest known switch
+    /// block such that the blocks' timestamps cover a duration of at least the max TTL for deploys
+    /// (a chainspec setting).
+    ///
+    /// If storage doesn't hold enough blocks to cover the specified duration, it will still return
+    /// the highest contiguous segment starting at the highest switch block which it does hold.
+    pub(crate) fn read_blocks_for_replay_protection(
         &self,
-        timestamp: Timestamp,
     ) -> Result<Vec<Block>, FatalStorageError> {
         let mut txn = self
             .env
             .begin_ro_txn()
             .expect("Could not start read only transaction for lmdb");
+        let timestamp = match self.switch_block_era_id_index.keys().last() {
+            Some(era_id) => self
+                .get_switch_block_header_by_era_id(&mut txn, *era_id)?
+                .map(|switch_block| switch_block.timestamp().saturating_sub(self.max_ttl))
+                .unwrap_or_else(Timestamp::now),
+            None => Timestamp::now(),
```

Comment:
We could also just return an empty `Vec` in this case? We wouldn't use the returned `Vec` anyway.